### PR TITLE
[fern] Round-2: Multi-scale RFF (3 frequency bands) for tau_y/tau_z

### PR DESCRIPTION
**Pull request by morganmcg1**
# [fern] Round-2: Multi-scale RFF — 3 frequency bands for tau_y / tau_z

## Hypothesis

Round-1 showed that Random Fourier Feature (RFF) coordinate encoding with sigma=1.0 lifted `test_abupt` from 19.81 → **17.77** (−10.3%). Fern's own analysis identified the key weakness: the single sigma=1.0 works well for meter-scale surface coordinates (domain ~[−1, 4] m), but **overshoots for far-field volume coordinates** (domain ~[−40, 80] m) — high-frequency Fourier features saturate in that regime, providing near-zero gradient signal.

Meanwhile, the primary gap to yi's SOTA lies in `tau_y` (21.81 vs 19.49) and `tau_z` (23.54 vs 21.12) — wall-shear components that depend on fine-scale surface details **and** meso-scale pressure gradients. These two components live at different spatial scales simultaneously; a single sigma cannot serve both.

**Multi-scale RFF** concatenates multiple frequency bands in one forward pass, giving the model simultaneous access to:

| Band | Surface sigma | Surface scale | Volume sigma | Volume scale |
|------|--------------|---------------|-------------|--------------|
| Low (global shape) | 0.1 | ~10 m | 0.01 | ~100 m |
| Mid (meso transitions) | 1.0 | ~1 m | 0.1 | ~10 m |
| High (fine detail) | 10.0 | ~0.1 m | 1.0 | ~1 m |

This is motivated by spectral bias theory (Rahimi & Recht 2007; Tancik et al. NeurIPS 2020 "Fourier Features Let Networks Learn High Frequency Functions"): a network trained with a single-scale positional encoding cannot simultaneously represent both low- and high-frequency spatial variation. Concatenating multiple bands breaks this limitation without increasing model depth.

The hypothesis: multi-scale RFF will close at least 1–2 abupt points on the new 17.77 baseline, primarily through improvement in `tau_y` and `tau_z`.

---

## Current Baseline (PR #33 — tay SOTA)

| Metric | tay PR #33 | yi best | AB-UPT |
|---|---:|---:|---:|
| `abupt` | **17.77** | 15.82 | — |
| `surface_pressure` | 11.20 | 9.99 | 3.82 |
| `wall_shear` | 18.68 | 16.60 | 7.29 |
| `volume_pressure` | 16.13 | 14.21 | 6.08 |
| `tau_x` | 16.20 | 14.27 | 5.35 |
| `tau_y` | 21.81 | 19.49 | 3.65 |
| `tau_z` | 23.54 | 21.12 | 3.63 |

Beat **17.77** on `test_abupt` to set a new SOTA.

Expected outcome: **−1.0 to −2.0 abupt** → target range **~15.8–16.8**.

---

### 1. Add `MultiScaleRFF` to `model.py`

Locate the existing `RFFEncoding` class (around line 49). Immediately **after** it, add:

```python
class MultiScaleRFF(nn.Module):
    """Concatenate RFF encodings at multiple frequency scales.

    Args:
        in_features: spatial coordinate dimensionality (e.g. 3 for xyz).
        num_features: number of Fourier features per sigma (before *2).
        sigmas: list of length-scale parameters; one RFFEncoding per sigma.

    Output dim = 2 * num_features * len(sigmas)
    """
    def __init__(self, in_features: int, num_features: int, sigmas: list[float]):
        super().__init__()
        self.encoders = nn.ModuleList([
            RFFEncoding(in_features, num_features, sigma=s)
            for s in sigmas
        ])
        self.out_features = 2 * num_features * len(sigmas)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # x: (..., in_features)  →  (..., out_features)
        return torch.cat([enc(x) for enc in self.encoders], dim=-1)
```

### 2. Add CLI arguments to `train.py`

Find the argument-parser block and add three new arguments (keep `--rff-sigma` for backward compatibility):

```python
parser.add_argument(
    "--rff-sigmas",
    type=str,
    default=None,
    help="Comma-separated list of RFF sigma values (overrides --rff-sigma when >1 value). "
         "E.g. '0.1,1.0,10.0'. Each sigma gets --rff-num-features features.",
)
parser.add_argument(
    "--rff-volume-sigmas",
    type=str,
    default=None,
    help="Optional comma-separated sigma list for volume coords (defaults to --rff-sigmas). "
         "E.g. '0.01,0.1,1.0'.",
)
```

### 3. Wire `MultiScaleRFF` into the model construction

In the block where `RFFEncoding` is instantiated for the surface / volume encoders, apply the following logic:

```python
def _parse_sigmas(sigma_str: str | None, fallback: float) -> list[float]:
    if sigma_str is not None:
        return [float(s.strip()) for s in sigma_str.split(",")]
    return [fallback]

surface_sigmas = _parse_sigmas(args.rff_sigmas, args.rff_sigma)
volume_sigmas  = _parse_sigmas(args.rff_volume_sigmas or args.rff_sigmas, args.rff_sigma)

if args.rff_num_features > 0:
    if len(surface_sigmas) > 1:
        surface_coord_encoder = MultiScaleRFF(3, args.rff_num_features, surface_sigmas)
    else:
        surface_coord_encoder = RFFEncoding(3, args.rff_num_features, sigma=surface_sigmas[0])

    if len(volume_sigmas) > 1:
        volume_coord_encoder = MultiScaleRFF(3, args.rff_num_features, volume_sigmas)
    else:
        volume_coord_encoder = RFFEncoding(3, args.rff_num_features, sigma=volume_sigmas[0])
```

**Important:** The output dimension of `MultiScaleRFF` is `2 * num_features * n_sigmas`. With `--rff-num-features 32` and 3 sigmas, each encoder outputs **192** features (vs 64 in Round-1). Verify that the downstream linear projection that consumes the coord encoding is either (a) dynamically sized from `encoder.out_features`, or (b) updated to accept 192. Check how `RFFEncoding.out_features` is currently consumed and apply the same pattern for `MultiScaleRFF.out_features`.

### 4. Run the experiment

Use exactly this command (single arm — do not sweep; concatenation is the entire hypothesis):

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --rff-num-features 32 \
  --rff-sigmas "0.1,1.0,10.0" \
  --rff-volume-sigmas "0.01,0.1,1.0" \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms --no-compile-model \
  --wandb-group fern-multiscale-rff
```

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

---

## Results Template

Please report results in a PR comment using the following format:

```
## Results

| Metric | Round-1 baseline (PR #33) | This run | Delta |
|---|---:|---:|---:|
| `test_abupt` | 17.77 | | |
| `surface_pressure` | 11.20 | | |
| `wall_shear` | 18.68 | | |
| `volume_pressure` | 16.13 | | |
| `tau_x` | 16.20 | | |
| `tau_y` | 21.81 | | |
| `tau_z` | 23.54 | | |

W&B run: <link>
Epoch of best checkpoint: 

## Analysis
<what improved, what didn't, suggested follow-ups>
```

---

## Notes

- **Single arm only** this round. Run the full 3-sigma concatenation directly — do not run an ablation sweep (that comes in Round 3 if this wins).
- `--rff-sigma` (singular) is still required in the code for backward compat; leave it in place, default `1.0`.
- If you hit a shape mismatch at the linear projection, print `encoder.out_features` and compare against the model's expected input dim. The fix is almost always to make the projection layer read from `encoder.out_features` rather than a hard-coded constant.
- If training diverges (loss spikes), try dropping `--lr` to `2e-5` before anything else.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern:

## Results

| Metric | Round-1 baseline (PR #33) | This run | Delta |
|---|---:|---:|---:|
| `test_abupt` | 17.77 | **17.86** | **+0.09** |
| `surface_pressure` | 11.20 | 11.28 | +0.08 |
| `wall_shear` | 18.68 | 18.69 | +0.01 |
| `volume_pressure` | 16.13 | 16.30 | +0.17 |
| `tau_x` | 16.20 | 16.14 | −0.06 |
| `tau_y` | 21.81 | 21.84 | +0.03 |
| `tau_z` | 23.54 | 23.73 | +0.19 |

W&B run: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/5bx6zsio
Group: `fern-multiscale-rff`
Epoch of best checkpoint: **9** (epoch-9 was a partial epoch, force-validated by the 270 min train timeout)
Best `full_val_primary/abupt_axis_mean_rel_l2_pct` = **17.0207** (Round-1: 17.0573)
Peak GPU memory: ~67.8 GiB / 96 GiB
Wall-clock runtime: 285 min (timeout-bounded)
DDP world size: 8 × RTX PRO 6000 Black

### Side-by-side full-val (best checkpoint, both at epoch 9)

| Metric | Round-1 (u43lik5d) | Multi-scale (5bx6zsio) | Δ |
|---|---:|---:|---:|
| `abupt` | 17.057 | 17.021 | −0.04 |
| `surface_pressure` | 11.638 | 11.736 | +0.10 |
| `wall_shear` | 19.167 | 19.186 | +0.02 |
| `volume_pressure` | 9.837 | **9.402** | **−0.43** |
| `tau_x` | 16.508 | 16.483 | −0.03 |
| `tau_y` | 22.465 | 22.479 | +0.01 |
| `tau_z` | 24.838 | 25.003 | +0.16 |

### Command used

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --wandb-name fern/round2-multiscale-rff-3band \
  --wandb-group fern-multiscale-rff \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --rff-num-features 32 \
  --rff-sigmas "0.1,1.0,10.0" --rff-volume-sigmas "0.01,0.1,1.0" \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms --no-compile-model
```

## Analysis

**Null result on the primary metric.** Multi-scale RFF (3 bands per modality) did NOT meaningfully improve `test_abupt`: 17.86 vs Round-1's 17.77 (+0.09, well inside noise). The hypothesis predicted −1 to −2 points of abupt improvement primarily through `tau_y` and `tau_z`; we got essentially flat on every wall-shear axis and slightly worse on `tau_z`. The 17.77 single-scale RFF baseline still holds.

**The one place multi-scale did help is `volume_pressure` on the *validation* split** — full_val volume_p dropped 9.84 → 9.40 (−0.43), about 4% relative. This is consistent with the spectral-bias prediction: the volume-coord bands `(0.01, 0.1, 1.0)` give the network access to far-field structure that single-sigma=1.0 saturates on. However, **this val improvement did not transfer to test** — `test_primary/volume_pressure_rel_l2_pct` actually got slightly worse (16.30 vs 16.13). So either the val/test split has heterogeneous far-field statistics, or multi-scale is overfitting the volume bands to validation.

**Why the surface side didn't move:**
- Surface-coord domain is ~[−1, 4] m; single sigma=1.0 already covers fine and meso scales for that range. Adding sigma=0.1 (low band) and sigma=10.0 (high band) provides little new information.
- The added 192-dim coord features (vs 64 in Round-1) didn't help training-signal richness — the surface projection layer just sees more redundant features.

**Compute cost.**
- Per-step throughput dropped from ~1.7 it/s (Round-1) to 1.49 it/s — ~14% slower steps.
- Per-epoch wall time grew from ~1064 s (Round-1 median) to ~1842 s (this run, ~73% slower).
- Most of that is the larger linear projection at the input (192→512 vs 64→512) and ~3× the RFF compute. The longer epoch meant within the 270-min train budget we got 8 full epochs + 1 partial vs Round-1's 9 full epochs.
- We were not training-loss-limited (train_loss 0.0799 vs Round-1 0.0794, basically tied) — the throughput hit just compresses how many gradient updates we accumulate.

**Honest read.** Multi-scale RFF as specified is essentially **parameter-equivalent to single-scale on this dataset** — the surface network does not benefit at all, and the volume gains do not generalize to test. This does not invalidate the spectral-bias hypothesis broadly; it just says that for DrivAerML at the current Transolver scale, sigma=1.0 + simple RFF is already in a local minimum of the multi-scale landscape.

## Suggested follow-ups

1. **Volume-only multi-scale.** Keep surface RFF single-sigma=1.0 (Round-1 winner) but add multi-scale only on volume coords with small sigmas `(0.01, 0.1, 1.0)`. Cuts per-step cost roughly in half vs this run, and isolates the only place multi-scale helped val.
2. **Reduce `rff_num_features` to 16** when using 3 bands, so total feature dim matches Round-1 (16 × 2 × 3 = 96 vs Round-1's 64). Removes the throughput penalty so we can fit ~13 epochs in the same budget.
3. **Ablation on volume sigmas.** Try volume sigmas `(0.01)` alone vs `(0.1)` alone vs `(1.0)` alone (Round-1 baseline) to figure out which band carries the −0.43 val volume_p improvement. The far-field 0.01 band is the leading suspect.
4. **Anti-overfitting on volume_p.** The val→test gap on volume (−0.43 val, +0.17 test) suggests the fancier far-field encoding overfits on 34 val cars. Consider stronger weight decay or feature-dropout on the volume RFF specifically.
5. **More epochs.** The Round-1 trajectory was still descending at epoch 9; we were similarly still descending here. If wall-clock budget allows 12-15 epochs, the multi-scale val curve might continue past the single-scale curve. Worth retrying once `torch.compile` is fixed (BASELINE.md notes that bug; would lift throughput ~2×).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #43 because ADVISOR: Closing this PR — clean null result, but the baseline has shifted.

## Outcome vs current SOTA

Current SOTA after this PR was assigned: **PR #39 Lion test_abupt = 15.43** (merged 2026-04-29 22:05 UTC).

Multi-scale RFF result:
- `test_abupt` = 17.86 → **+15.7% regression vs SOTA** (well past the 5% close-threshold)
- vs your own Round-1 baseline (PR #33, 17.77): +0.09, within noise

Your honest read of the result is exactly right. Multi-scale RFF on the standard surface-coord domain `[−1,4] m` is parameter-redundant with sigma=1.0 single-scale; the surface network gets nothing new, and the val gain on `volume_pressure` (9.84→9.40, −0.43) didn't transfer to test. Two clean takeaways:

1. **Single-scale sigma=1.0 RFF is already at the local optimum** for this Transolver scale on DrivAerML.
2. **Spectral-bias hypothesis is not invalidated** — it just doesn't have headroom on the surface side at the current backbone size.

## Why I'm closing rather than sending back

Multi-scale RFF on top of AdamW (the base of this experiment) won't beat Lion-base. The follow-ups you proposed (volume-only multi-scale, narrower bands, more epochs) are each interesting but each is its own hypothesis — they shouldn't share this PR. If a strong volume-only multi-scale result emerges later, we can reopen the thread.

## What I'm assigning you next

Single-delta from the new Lion SOTA, addressing the actual binding constraint on test_abupt: **per-axis loss weighting for `tau_y` and `tau_z`**, the two worst axes (19.58 / 20.40, both ~5×+ gap to AB-UPT). New PR coming.

Excellent work — the rigor and self-skepticism in this writeup is exactly what we want..
